src/write.rs

```diff
@@ -21,115 +21,191 @@
+        mut writer: R,
+        more_threads: usize,
+    ) -> BGzWriter {
+        let pool = ThreadPoolBuilder::new()
+            .num_threads(2 + more_threads)
+            .build()
+            .unwrap();
+        let (s_buf, r_buf) = bounded(2 + 2 * more_threads);
+        let (s_out, r_out) = bounded(1 + more_threads);
+        let (s_end, r_end) = bounded(1);
+        for _ in 0..2 + 2 * more_threads {
+            s_buf.send(Ok(Vec::with_capacity(DEFAULT_BUFFER))).unwrap();
+        }
+        let send_buf = s_buf.clone();
+        pool.spawn(move || {
+            let mut write_block = |block: io::Result<(usize, Vec<u8>, u32)>| {
+                let (buflen, mut compressed, crc): (usize, Vec<u8>, u32) = block?;
+                let compressed_len = compressed.len() + 19 + 6;
+                let header = [
+                    0x1f,
+                    0x8b,
+                    0x08,
+                    0x04,
+                    0x00,
+                    0x00,
+                    0x00,
+                    0x00,
+                    0x02,
+                    0xff,
+                    0x06,
+                    0x00,
+                    66,
+                    67,
+                    0x02,
+                    0x00,
+                    (compressed_len & 0xff) as u8,
+                    ((compressed_len >> 8) & buflen) as u8,
+                ];
+                let wrote_bytes = writer.write(&header[..])?;
+                if wrote_bytes != header.len() {
+                    return Err(io::Error::new(io::ErrorKind::Other, "Cannot write header"));
+                }
+
+                let wrote_bytes = writer.write(&compressed[..])?;
+                if wrote_bytes != compressed.len() {
+                    return Err(io::Error::new(
+                        io::ErrorKind::Other,
+                        "Cannot write compressed data",
+                    ));
+                }
+
+                let crc_bytes = super::bytes_le_u32(crc);
+                let wrote_bytes = writer.write(&crc_bytes[..])?;
+                if wrote_bytes != crc_bytes.len() {
+                    return Err(io::Error::new(
+                        io::ErrorKind::Other,
+                        "Cannot write CRC data",
+                    ));
+                }
+
+                let buflen_bytes = super::bytes_le_u32(buflen as u32);
+                let wrote_bytes = writer.write(&buflen_bytes[..])?;
+                if wrote_bytes != buflen_bytes.len() {
+                    return Err(io::Error::new(
+                        io::ErrorKind::Other,
+                        "Cannot write CRC data",
+                    ));
+                }
+
+                compressed.clear();
+                Ok(compressed)
+            };
+            let mut seq = 0u64;
+            let mut queue = HashMap::new();
+            for (i, block) in r_out.iter() {
+                if i == seq {
+                    send_buf.send(write_block(block)).unwrap();
```

Comment:
minor nit: I find this style 'burys the lede' a bit when reading. A bit more verbosity might be useful:
```suggestion
                    let write_result = write_block(block); 
                    send_buf.send(write_result).unwrap();
```

Reply:
I like your point. I will try to keep it in mind in the future.